src/mms/MMS.cpp

```diff
@@ -0,0 +1,188 @@
+            const amrex::Real y = problo[1] + (j + 0.5) * dx[1];
+            const amrex::Real z = problo[2] + (k + 0.5) * dx[2];
+
+            den(i, j, k) = 1.0;
+            vel(i, j, k, 0) = masa_eval_3d_exact_u(x, y, z);
+            vel(i, j, k, 1) = masa_eval_3d_exact_v(x, y, z);
+            vel(i, j, k, 2) = masa_eval_3d_exact_w(x, y, z);
+        });
+    }
+    amrex::prefetchToDevice(velocity);
+    amrex::prefetchToDevice(density);
+}
+
+/** Fill the MMS source term.
+ */
+void MMS::fill_src()
+{
+    m_mms_vel_source.setVal(0.0);
+
+    const int nlevels = m_repo.num_active_levels();
+    for (int lev = 0; lev < nlevels; ++lev) {
+        const auto& dx = m_mesh.Geom(lev).CellSizeArray();
+        const auto& problo = m_mesh.Geom(lev).ProbLoArray();
+        auto& mms_src_term = m_mms_vel_source(lev);
+#ifdef _OPENMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+        for (amrex::MFIter mfi(mms_src_term, amrex::TilingIfNotGPU());
+             mfi.isValid(); ++mfi) {
+            const auto& bx = mfi.tilebox();
+            const auto& mms_src = mms_src_term.array(mfi);
+
+            amrex::LoopOnCpu(bx, [=](int i, int j, int k) noexcept {
+                const amrex::Real x = problo[0] + (i + 0.5) * dx[0];
+                const amrex::Real y = problo[1] + (j + 0.5) * dx[1];
+                const amrex::Real z = problo[2] + (k + 0.5) * dx[2];
+
+                mms_src(i, j, k, 0) = masa_eval_3d_source_u(x, y, z);
+                mms_src(i, j, k, 1) = masa_eval_3d_source_v(x, y, z);
+                mms_src(i, j, k, 2) = masa_eval_3d_source_w(x, y, z);
+            });
+        }
+        amrex::prefetchToDevice(mms_src_term);
+    }
+}
+
+void MMS::post_init_actions() { fill_src(); }
+
+void MMS::post_regrid_actions() { fill_src(); }
+
+void MMS::post_advance_work()
+{
+
+    amrex::Real u_mms_err = 0.0;
+    amrex::Real v_mms_err = 0.0;
+    amrex::Real w_mms_err = 0.0;
+
+    const int nlevels = m_repo.num_active_levels();
+    for (int lev = 0; lev < nlevels; ++lev) {
+
+        amrex::iMultiFab level_mask;
+        if (lev < nlevels - 1) {
+            level_mask = makeFineMask(
+                m_mesh.boxArray(lev), m_mesh.DistributionMap(lev),
+                m_mesh.boxArray(lev + 1), amrex::IntVect(2), 1, 0);
+        } else {
+            level_mask.define(
+                m_mesh.boxArray(lev), m_mesh.DistributionMap(lev), 1, 0,
+                amrex::MFInfo());
+            level_mask.setVal(1);
+        }
+
+        const auto& dx = m_mesh.Geom(lev).CellSizeArray();
+        const auto& problo = m_mesh.Geom(lev).ProbLoArray();
+        const amrex::Real cell_vol = dx[0] * dx[1] * dx[2];
+
+        using FuncDef = double (*)(double, double, double);
+        auto compute_error = [this, &lev, &level_mask, &problo, &dx,
+                              &cell_vol](int idx, FuncDef f) {
+            amrex::Real error = amrex::ReduceSum(
```

Comment:
We should revisit this reduce call. If the innermost function is going to run on CPU, launching reduction kernels on device don't make sense.

Reply:
Any suggestions for that @sayerhs ? I moved things around a bit...

Reply:
I think we can do this entirely on host instead of going through the `ReduceSum` kernel launch. That means using an accumulation variable, MFIter loops with `amrex::Loop` within that handles the per-rank logic and then doing MPI reductions on the variable. I think the `compute_error` function can go away :(

Reply:
Ok I took a stab at it. Let me know if you see ways to improve it.